### PR TITLE
Save

### PR DESCRIPTION
**Pull request by kevinlmadison**
Ok I'm expecting some backlash but I'll go ahead and create the request. This does in fact create a new save file with the serialized content and is loadable from a completely separate instance of the program.

### CODE REVIEW

**Comment by kevinlmadison** [comment]
Ok so I don't really know how to work with Travis yet, I clicked on the details link but I don't see where it's giving me an issue.

**Comment by pbateman828** [comment]
Yeah, rustfmt is saying your code isn't formated correctly in travis

**Comment by pbateman828** [comment]
Travis builds it and displays whats wrong within the black box. If you change it and push it to the branch, travis automatically builds it again.

File: src/io.rs [hunk 1]
```diff
@@ -3,7 +3,11 @@
 use std::fs::File;
 use std::path::Path;
 use protobuf::core::MessageStatic;
+use protobuf::CodedOutputStream;
```
Comment (pbateman828) [inline src/io.rs]: warning: unused import: `protobuf::CodedOutputStream`
 --> src/load.rs:7:5
  |
7 | use protobuf::CodedOutputStream;
  |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^
  |
  = note: #[warn(unused_imports)] on by default

File: src/io.rs [hunk 2]
```diff
@@ -3,7 +3,11 @@
...
+use protobuf::Message;
```
Comment (pbateman828) [inline src/io.rs]: --> src/load.rs:8:5
  |
8 | use protobuf::Message;
  |     ^^^^^^^^^^^^^^^^^

File: src/io.rs [hunk 3]
```diff
@@ -14,22 +18,36 @@ where
...
+                Err(_) => Err(String::from("Failed to load file.")),
+            }
+        }
         Err(_) => Err(String::from("Failed to open file")),
     };
 }
 
+pub fn write_protobuf<I: Interfaceable>(
+    src: &mut I,
+    mut game: Game,
```
Comment (pbateman828) [inline src/io.rs]: variable does not need to be mutable
  --> src/io.rs:33:5
   |
33 |     mut game: Game,
   |     ---^^^^^
   |     |
   |     help: remove this `mut`
   |
   = note: #[warn(unused_mut)] on by default

File: src/load.rs [hunk 4]
```diff
@@ -24,13 +27,42 @@ pub fn new<I: Interfaceable>(
...
-    let path = src.prompt();
+pub fn save<I: Interfaceable>(
+    src: &mut I,
+    game: &Game,
+) -> File {
 
-    File::create(&Path::new(&path)).unwrap();
-    return Ok(game);
+    let mut path;
+    let mut file;
```
Comment (pbateman828) [inline src/load.rs]: warning: variable does not need to be mutable
  --> src/load.rs:39:9
   |
39 |     let mut file;
   |         ---^^^^^
   |         |
   |         help: remove this `mut`

**Comment by kevinlmadison** [comment]
Oh ok it doesn't fix it for you it just shows you where it went wrong?

File: src/io.rs [hunk 5]
```diff
@@ -14,22 +18,36 @@ where
...
+) -> Result<Game, String> {
+
+    let mut file = save(src, &game);
+    let mut cos = CodedOutputStream::new(&mut file);
+    game.write_to(&mut cos);
+    cos.flush();
```
Comment (pbateman828) [inline src/io.rs]: warning: unused `std::result::Result` which must be used
  --> src/io.rs:38:5
   |
38 |     game.write_to(&mut cos);
   |     ^^^^^^^^^^^^^^^^^^^^^^^^
   |
   = note: #[warn(unused_must_use)] on by default

adding .expect() may work

**Review by pbateman828 — CHANGES_REQUESTED** [review]

**Comment by pbateman828** [comment]
Yeah, it just shows you. Also, I just created a save file and loaded that same file in another instance of the game. Pretty swag

File: src/io.rs (hunk 5 above)
Comment (pbateman828) [reply src/io.rs]: game.write_to(&mut cos).expect("Something went wrong");

works

**Comment by kevinlmadison** [comment]
Can I squash on github or do I need to rebase locally before I push again?

**Comment by pbateman828** [comment]
Ive always squashed locally then pushed to the branch I wanna merge with master

**Comment by pbateman828** [comment]
Just add +use protobuf::Message;⏎

**Review by pbateman828 — APPROVED** [review]

File: src/io.rs.bk [hunk 6]
```diff
@@ -0,0 +1,80 @@
+//! Operations that interact with the file system.
+
```
Comment (dmadisetti) [inline src/io.rs.bk]: Delete this file

File: src/io.rs [hunk 7]
```diff
@@ -1,8 +1,12 @@
 //! Operations that interact with the file system.
 
+use load::save;
```
Comment (dmadisetti) [inline src/io.rs]: Alphabetize

File: src/load.rs [hunk 8]
```diff
@@ -4,6 +4,7 @@ use std::fs::File;
 use std::path::Path;
 use protos::master::Game;
 use screen::Interfaceable;
+use io::write_protobuf;
```
Comment (dmadisetti) [inline src/load.rs]: Alphabetize

File: src/load.rs [hunk 9]
```diff
@@ -24,13 +25,40 @@ pub fn new<I: Interfaceable>(
...
 
-    File::create(&Path::new(&path)).unwrap();
-    return Ok(game);
+        path = src.prompt();
+
+        if Path::new(&path).exists() {
+            src.print(
+                "This file already exists, would you like to overwrite it?",
+            );
+            src.print("1 - Yes");
```
Comment (dmadisetti) [inline src/load.rs]: I think this could be a utility function `fn confirm` under interfacable

File: src/io.rs [hunk 10]
```diff
@@ -22,14 +26,25 @@ where
...
+) -> Result<Game, String> {
+    let mut file = save(src, &game);
+    let mut cos = CodedOutputStream::new(&mut file);
+    game.write_to(&mut cos).unwrap();
+    cos.flush().unwrap();
+    return Ok(game);
+}
+
 pub fn prompt_path<F, M: MessageStatic, I: Interfaceable>(
     src: &mut I,
```
Comment (dmadisetti) [inline src/io.rs:49]: I think we could probably rework this into screen.rs and call from `fn save` as well

**Review by dmadisetti — APPROVED** [review]
Fix the minor nits before you submit (alphabetize, delete etc.. no big reworks). It gets the major parts. I left comments showing a possibility for refactor, but I'll grab that myself this afternoon

File: src/io.rs [hunk 11]
```diff
@@ -22,14 +26,25 @@ where
     };
 }
 
+pub fn write_protobuf<I: Interfaceable>(
+    src: &mut I,
+    game: Game,
+) -> Result<Game, String> {
+    let mut file = save(src, &game);
+    let mut cos = CodedOutputStream::new(&mut file);
+    game.write_to(&mut cos).unwrap();
```
Comment (dmadisetti) [inline src/io.rs]: Handle this error unwrap to prevent panic, and return Err(problem) gracefully

**Review by dmadisetti — CHANGES_REQUESTED** [review]
Woops one more thing you should grab before submitting

File: src/io.rs.bk (hunk 6 above)
Comment (kevinlmadison) [reply src/io.rs.bk]: will do, I'm guessing that came from using rustfmt.

**Comment by kevinlmadison** [comment]
I don't think I squashed correctly...

**Comment by pbateman828** [comment]
? how so

File: src/io.rs (hunk 10 above)
Comment (kevinlmadison) [reply src/io.rs:49]: I did not do this but I did all the other things

**Review by dmadisetti — APPROVED** [review]

**Comment by dmadisetti** [comment]
Good enough for now. A few things I might touch up when I refactor

**Comment by kevinlmadison** [comment]
dank

**Pull request merged**